crates/turbo-tasks-memory/src/task.rs

```diff
@@ -1521,6 +1970,57 @@ impl Task {
     }
 }
 
+fn remove_collectible_from_scopes(
+    emitted: AutoSet<(TraitTypeId, RawVc)>,
+    unemitted: AutoSet<(TraitTypeId, RawVc)>,
+    task_scopes: &TaskScopes,
+    backend: &MemoryBackend,
+    turbo_tasks: &dyn TurboTasksBackendApi,
+) {
+    task_scopes.iter().for_each(|id| {
+        backend.with_scope(id, |scope| {
+            let mut tasks = AutoSet::new();
+            {
+                let mut state = scope.state.lock();
+                emitted
+                    .iter()
+                    .filter_map(|(trait_id, collectible)| {
+                        state.remove_collectible(*trait_id, *collectible)
+                    })
+                    .for_each(|e| tasks.extend(e.notify));
+
+                unemitted
+                    .iter()
+                    .filter_map(|(trait_id, collectible)| {
+                        state.add_collectible(*trait_id, *collectible)
+                    })
+                    .for_each(|e| tasks.extend(e.notify));
```

Comment:
Perhaps this could also be
```suggestion
                tasks.extend(unemitted
                    .iter()
                    .filter_map(|(trait_id, collectible)| {
                        state.add_collectible(*trait_id, *collectible)
                    })
                    .flat_map(|e| e.notify.into_iter()));
```